src/java.base/macosx/classes/sun/nio/fs/BsdFileSystem.java

```diff
@@ -71,13 +79,89 @@ public Set<String> supportedFileAttributeViews() {
         return SupportedFileFileAttributeViewsHolder.supportedFileAttributeViews;
     }
 
+    /**
+     * Clones the file whose path name is {@code src} to that whose path
+     * name is {@code dst} using the {@code clonefile} system call.
+     *
+     * @param src the path of the source file
+     * @param dst the path of the destination file (clone)
+     * @param followLinks whether to follow links
+     *
+     * @return 0 on success, IOStatus.UNSUPPORTED_CASE if the call does not work
+     *         with the given parameters, or IOStatus.UNSUPPORTED if cloning is
+     *         not supported on this platform
+     */
+    private int clone(UnixPath src, UnixPath dst, boolean followLinks)
+        throws IOException
+    {
+        // Do not clone unless both files are on the same volume and that
+        // volume is indicated as supporting file cloning
+        BsdFileStore bfs = (BsdFileStore)provider().getFileStore(src);
+        if (!bfs.equals(provider().getFileStore(dst.getParent())) ||
+            !bfs.supportsCloning())
+            return IOStatus.UNSUPPORTED_CASE;
+
+        int flags = followLinks ? 0 : CLONE_NOFOLLOW;
+        try {
+            BsdNativeDispatcher.clonefile(src, dst, flags);
+        } catch (UnixException x) {
+            switch (x.errno()) {
+                case ENOTSUP: // cloning not supported by filesystem
+                    return IOStatus.UNSUPPORTED;
```

Comment:
It might be that `ENOTSUP` should instead result in `UNSUPPORTED_CASE` and that the static variable `cloneFileNotSupported` is unnecessary.

Reply:
Yes, ENOTSUP depends on the parameters to I think it maps to UNSUPPORTED_CASE. Given that it has to handle ENOTSUP then it may be that the check for the VOL_CAP_INT_CLONE feature isn't needed. You may have to compare to the performance of getattrlist to see if it helps when the file systems don't support clone.

Reply:
ENOTSUP is mapped to UNSUPPORTED_CASE in f51cd186fe624d03ce7c6a4f49140998e2d48322, which also removes getattrlist(). Benchmark testing in Java was inconclusive, but testing C code showed an improvement of only 1 microsecond for checking VOL_CAP_INT_CLONE on a volume which does not supporting cloning as opposed to just letting clonefile() set ENOTSUP. The extra code for getattrlist() does not seem to be worth it.